compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/FlattenStringConcatenationLowering.kt

```diff
@@ -76,21 +77,42 @@ class FlattenStringConcatenationLowering(val context: CommonBackendContext) : Fi
             KotlinBuiltIns.BUILT_INS_PACKAGE_FQ_NAME,
             KotlinBuiltIns.FQ_NAMES.string.toSafe()
         )
-        private val PLUS_NAME = Name.identifier("plus")
 
-        /** @return true if the given expression is a [IrStringConcatenation] or [String.plus] [IrCall]. */
+        /** @return true if the given expression is a [IrStringConcatenation], or an [IrCall] to [toString] or [String.plus]. */
         private fun isStringConcatenationExpression(expression: IrExpression): Boolean {
             return when (expression) {
                 is IrStringConcatenation -> true
                 is IrCall -> {
+                    if (expression.superQualifierSymbol != null)
+                        return false
+
                     val function = expression.symbol.owner
-                    val receiver = expression.dispatchReceiver ?: expression.extensionReceiver
-                    receiver != null &&
-                            receiver.type.isStringClassType() &&
-                            expression.type.isStringClassType() &&
-                            expression.valueArgumentsCount == 1 &&
-                            function.name == PLUS_NAME &&
-                            function.fqNameWhenAvailable?.parent() in PARENT_NAMES
+                    when (function.name.asString()) {
+                        "toString" -> {
+                            // Check that this is a call to method Any.toString or to the extension function
+                            // Any?.toString defined in the kotlin package.
+                            expression.valueArgumentsCount == 0
+                                    && ((expression.dispatchReceiver != null
+                                    && (expression.type.isAny() || (function as IrSimpleFunction).overriddenSymbols.isNotEmpty()))
```

Comment:
Why `isAny()` and not `isString()` here?

Also, I think we need to check that extension receiver is null in this case. Overall, consider splitting this to two functions for easier reading

Reply:
That was supposed to be a test on the receiver parameter, fixed! I also split the function to make it easier to read.